File: code/packages/dart/single-layer-network/test/single_layer_network_test.dart
import 'package:test/test.dart';
import 'package:coding_adventures_single_layer_network/coding_adventures_single_layer_network.dart';

void main() {
  test('one epoch exposes matrix gradients', () {
    final step = trainOneEpochWithMatrices(
      [
        [1.0, 2.0],
      ],
      [
        [3.0, 5.0],
      ],
      [
        [0.0, 0.0],
        [0.0, 0.0],
      ],
      [0.0, 0.0],
      0.1,
    );

    expect(step.weightGradients[0][0], closeTo(-3.0, 1e-6));
    expect(step.weightGradients[1][1], closeTo(-10.0, 1e-6));
    expect(step.nextWeights[0][0], closeTo(0.3, 1e-6));
    expect(step.nextWeights[1][1], closeTo(1.0, 1e-6));
  });

  test('fit learns m inputs to n outputs', () {
    final model = SingleLayerNetwork(3, 2);
    final history = model.fit(
      [
        [0.0, 0.0, 1.0],
        [1.0, 2.0, 1.0],
        [2.0, 1.0, 1.0],
      ],
      [
        [1.0, -1.0],
        [3.0, 2.0],
        [4.0, 1.0],
      ],
      learningRate: 0.05,
      epochs: 500,
    );
    expect(history.last.loss, lessThan(history.first.loss));
    expect(model.predict([
      [1.0, 1.0, 1.0],
    ]).first.length, equals(2));
  });
}

File: code/packages/elixir/single_layer_network/test/single_layer_network_test.exs
defmodule CodingAdventures.SingleLayerNetworkTest do
  use ExUnit.Case

  alias CodingAdventures.SingleLayerNetwork

  defp near(actual, expected), do: assert(abs(actual - expected) <= 1.0e-6)

  test "one epoch exposes matrix gradients" do
    step =
      SingleLayerNetwork.train_one_epoch_with_matrices(
        [[1.0, 2.0]],
        [[3.0, 5.0]],
        [[0.0, 0.0], [0.0, 0.0]],
        [0.0, 0.0],
        0.1
      )

    assert step.predictions == [[0.0, 0.0]]
    assert step.errors == [[-3.0, -5.0]]
    assert step.weight_gradients == [[-3.0, -5.0], [-6.0, -10.0]]
    near(Enum.at(Enum.at(step.next_weights, 0), 0), 0.3)
    near(Enum.at(Enum.at(step.next_weights, 1), 1), 1.0)
  end

  test "fit learns m inputs to n outputs" do
    {model, history} =
      SingleLayerNetwork.new(3, 2)
      |> SingleLayerNetwork.fit(
        [[0.0, 0.0, 1.0], [1.0, 2.0, 1.0], [2.0, 1.0, 1.0]],
        [[1.0, -1.0], [3.0, 2.0], [4.0, 1.0]],
        0.05,
        500
      )

    assert List.last(history).loss < hd(history).loss
    assert length(hd(SingleLayerNetwork.predict(model, [[1.0, 1.0, 1.0]]))) == 2
  end
end
